### ISSUE TITLE
In a Silent Update (AutomaticallyDownloadUpdates), how could I know if Download is finished ?

### ISSUE THREAD

**Issue by prasha6** [sparkle-project/Sparkle#845]
Hello,

In a Silent Update (AutomaticDownloadUpdate - no sparkle UI) - Is there a way to know when the download is complete.

Basically, i want to set a few properties in my program, once the download is complete.

I don't think SUUpdaterDelegate's willInstallUpdate will work, because it is called right before installing an update. I want something that could be called right after the download is complete.

**Comment by pornel** [comment]
What properties do you want to set?

**Comment by prasha6** [comment]
Hi pornel,

Properties specific to my app. I have a few settings in my app, that I need to update after the download is complete. 

Few logic in my code will leverage on this app settings. For Example: I don't want my app to do a few things, if an update is downloaded.

**Comment by pornel** [comment]
We do have more delegate methods: https://github.com/sparkle-project/Sparkle/search?utf8=✓&q=delegate+respondsToSelector

but don't know if any of them will suit you, because it's not clear to me why you want to do something with your app specifically after download - you're explaining what, not why: http://xyproblem.info/

If you're trying to prepare your app to be updated, keep in mind that download doesn't mean much. Update being downloaded doesn't guarantee it'll be installed. For example if user shuts the system down, we will discard the downloaded file and won't install it.

Similarly, the install can be postponed or aborted if we show permission prompt. And the installation can be cancelled if it needs admin permissions, and user refuses to give them.

**Comment by prasha6** [comment]
Here is more background:

My app constantly checks if it needs a silent update; when the main window of the app is activated.

Lets call this method "NeedsSilentUpdate".
NeedsSilentUpdate would get the "downloadedVersion" property from app settings and compare it against the version in MyCustom.xml. If the "downloadedVersion" is greater than or equal to the MyCustom.xml, then the app doesn't need a silent update and "NeedSilentUpdate" can return false.

Note: MyCustom.xml is different from appcast.xml. I am using MyCustom.xml to check if there is a valid update, and I am leveraging the appcast.xml to just download the update.

Also, only when "NeedsSilentUpdate" returns true, I instantiate SUUpdater, set the feedURL to the appcast.xml location, and call CheckforUpdatesinBackground().

When the app downloads a silent update, I would set a property "downloadedversion"  in the app settings equal to the downloaded version.

I am doing this because:
1. I want to know if the update is already downloaded, when the call to NeedSilentUpdate is made.
2. What if there is a valid update, but sparkle fails to download an update (due to connectivity issues). 
3. There are chances the user doesn't quit an app for a long time, and multiple updates could be distributed in that time. So, I check for updates every 24 hours.

**Comment by prasha6** [comment]
In addition, since I am calling SUUpdater.checkforupdatesinbackground() programmatically, I need to know when I should be calling it. The 'NeedsSilentUpdate' tells me when I should be calling SUUpdater's checkforupdatesinbackground().

**Comment by pornel** [comment]
Do you do your own checking, because you want some updates to be silent, and some other updates to show Ui?

**Comment by prasha6** [comment]
Yes, that's right.

I figured that I could take advantage of "UpdaterWillDownloadUpdate" and "UpdaterFailedToDownloadUpdate" methods in the SUUpdaterDelegate class. I could set my properties in the "UpdaterWillDownloadUpdate" and clear the properties if the updater fails to download the update in the "UpdaterFailedToDownloadUpdate".

Do you think this is a viable approach ?

Thanks,
Prashanth

**Comment by pornel** [comment]
I think we may just add a tag to the appcast, or a delegate callback to say that an update is supposed to be silent.
